src/raftstore/store/worker/apply.rs

```diff
@@ -960,26 +958,31 @@ impl ApplyDelegate {
                     .with_label_values(&["add_peer", "all"])
                     .inc();
 
-                if exists {
-                    error!(
-                        "{} can't add duplicated peer {:?} to region {:?}",
-                        self.tag, peer, self.region
-                    );
-                    return Err(box_err!(
-                        "can't add duplicated peer {:?} to region {:?}",
-                        peer,
-                        self.region
-                    ));
+                let mut exists = false;
+                if let Some(p) = util::find_peer_mut(&mut region, store_id) {
+                    exists = true;
+                    if !p.get_is_learner() || p.get_id() != peer.get_id() {
+                        error!(
+                            "{} can't add duplicated peer {:?} to region {:?}",
+                            self.tag, peer, self.region
+                        );
+                        return Err(box_err!(
+                            "can't add duplicated peer {:?} to region {:?}",
+                            peer,
+                            self.region
+                        ));
+                    } else {
+                        p.set_is_learner(false);
+                    }
+                }
```

Comment:
I believe the pattern of

```rust
if let ... {
  // ...
} else {
  // ...
}
```

If preferred over

```rust
let mut flag = false;
if let ... {
  flag = true;
  // ...
}
if flag {
  // ...
}
```

Reply:
yes, but `if let` will borrow something so that we can't borrow it again in `else` branch. It's a lifetime issue. So I will comment it.

Reply:
Ah! This makes sense. Thanks for clarifying.